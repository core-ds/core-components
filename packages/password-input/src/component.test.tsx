import React, { useState, FC } from 'react';
import { render, fireEvent } from '@testing-library/react';

import { PasswordInput, PasswordInputProps } from '.';

const ControlledPasswordInput: FC<PasswordInputProps> = (props) => {
    const [passwordVisible, setPasswordVisible] = useState(false);

    const handlePasswordVisibleChange = (visible: boolean) => {
        setPasswordVisible(visible);
    };

    return (
        <PasswordInput
            {...props}
            passwordVisible={passwordVisible}
            onPasswordVisibleChange={handlePasswordVisibleChange}
        />
    );
};

const isPasswordHidden = (input: HTMLInputElement) => input.getAttribute('type') === 'password';
const isPasswordVisible = (input: HTMLInputElement) => input.getAttribute('type') === 'text';

describe('PasswordInput', () => {
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation((query) => ({
            matches: true,
            media: query,
            onchange: null,
            addListener: jest.fn(), // Deprecated
            removeListener: jest.fn(), // Deprecated
            addEventListener: jest.fn(),
            removeEventListener: jest.fn(),
            dispatchEvent: jest.fn(),
        })),
    });

    const dataTestId = 'test-id';

    describe('snapshots tests', () => {
        it('should match snapshot', () => {
            expect(render(<PasswordInput />).baseElement).toMatchSnapshot();
        });
    });

    describe('interactions tests', () => {
        it('should switch password visibility (uncontrolled)', () => {
            const { baseElement, getByTestId } = render(<PasswordInput dataTestId={dataTestId} />);

            const button = baseElement.querySelector('button') as HTMLButtonElement;
            const input = getByTestId(dataTestId) as HTMLInputElement;

            expect(isPasswordHidden(input)).toBe(true);

            fireEvent.click(button);

            expect(isPasswordVisible(input)).toBe(true);

            fireEvent.click(button);

            expect(isPasswordHidden(input)).toBe(true);
        });

        it('should switch password visibility (controlled)', () => {
            const { baseElement, getByTestId } = render(
                <ControlledPasswordInput dataTestId={dataTestId} />,
            );

            const button = baseElement.querySelector('button') as HTMLButtonElement;
            const input = getByTestId(dataTestId) as HTMLInputElement;

            expect(isPasswordHidden(input)).toBe(true);

            fireEvent.click(button);

            expect(isPasswordVisible(input)).toBe(true);

            fireEvent.click(button);

            expect(isPasswordHidden(input)).toBe(true);
        });
    });

    describe('check ToolTipHint', () => {
        let getByRole: ReturnType<typeof render>['getByRole'];
        let rerender: ReturnType<typeof render>['rerender'];
        let onPasswordVisibleChange: jest.Mock;

        describe('SUCCESS_CASES', () => {
            beforeEach(() => {
                ({ getByRole } = render(<PasswordInput />));
            });

            it('should initially display tooltip "Показать"', () => {
                expect(getByRole('button')).toHaveAttribute('title', 'Показать');
            });

            it('should display tooltip "Скрыть" after one click', () => {
                const eyeButton = getByRole('button');
                fireEvent.click(eyeButton);

                expect(eyeButton).toHaveAttribute('title', 'Скрыть');
            });

            it('should revert to tooltip "Показать" after two clicks', () => {
                const eyeButton = getByRole('button');
                fireEvent.click(eyeButton);
                fireEvent.click(eyeButton);

                expect(eyeButton).toHaveAttribute('title', 'Показать');
            });
        });

        describe('EDGE_CASES', () => {
            beforeEach(() => {
                onPasswordVisibleChange = jest.fn();
                ({ getByRole, rerender } = render(
                    <PasswordInput
                        passwordVisible={false}
                        onPasswordVisibleChange={onPasswordVisibleChange}
                    />,
                ));
            });

            it('should call onPasswordVisibleChange with true when clicked in controlled mode', () => {
                const eyeButton = getByRole('button');
                fireEvent.click(eyeButton);

                expect(onPasswordVisibleChange).toHaveBeenCalledWith(true);
            });

            it('should display tooltip "Скрыть" after re-rendering with passwordVisible true', () => {
                rerender(
                    <PasswordInput passwordVisible={true} onPasswordVisibleChange={jest.fn()} />,
                );

                expect(getByRole('button')).toHaveAttribute('title', 'Скрыть');
            });
        });

        describe('ERROR_CASES', () => {
            it('should be disabled', () => {
                let getByRole: ReturnType<typeof render>['getByRole'];
                ({ getByRole } = render(<PasswordInput disabled />));

                expect(getByRole('button')).toBeDisabled();
            });
        });
    });
});
