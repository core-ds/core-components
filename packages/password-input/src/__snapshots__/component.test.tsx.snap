// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordInput snapshots tests should match snapshot 1`] = `
<body>
  <div>
    <div
      class="component component size-48 size-48 size-48 hasRightAddons"
    >
      <div
        class="inner inner inner inner"
      >
        <div
          class="inputWrapper"
        >
          <div
            class="input size-48"
          >
            <input
              class="input size-48 input input"
              type="password"
              value=""
            />
          </div>
        </div>
        <div
          class="addons rightAddons addons"
        >
          <button
            class="component text size-48 hug size-48 component text component text cc-icon-button eye secondary component border-40"
            title="Показать введённые цифры"
            type="button"
          >
            <span
              class="label stretchText"
            >
              <span
                class="iconWrapper size-48 center"
              >
                <svg
                  class="icon"
                  fill="currentColor"
                  focusable="false"
                  height="24"
                  role="img"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <path
                    clip-rule="evenodd"
                    d="M22.015 18.584L5.432 2 4.018 3.413l2.78 2.78C4.252 7.412 2.177 9.484 1 12.003c1.914 4.09 6.19 7.002 11 7.002 2.164 0 4.22-.589 6.001-1.61l2.6 2.601 1.414-1.413zm-6.756-3.93l-1.573-1.573a2 2 0 11-2.763-2.763L9.35 8.745a4.2 4.2 0 105.91 5.91z"
                    fill-rule="evenodd"
                  />
                  <path
                    d="M23 12.004a11.952 11.952 0 01-1.813 2.778.1.1 0 01-.145.005L11.438 5.18a.098.098 0 01.065-.169c.165-.007.33-.01.497-.01 4.81 0 9.086 2.912 11 7.002z"
                  />
                </svg>
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
